pkg/storage/gcp/saver.go

```diff
@@ -0,0 +1,75 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/option"
+)
+
+// Storage implements the (github.com/gomods/pkg/storage).Saver interface
+type Storage struct {
+	bucket *storage.BucketHandle
+}
+
+// New returns a new Storage instance
+// authenticated using the provided ClientOptions for the associated bucket
+func New(bucketname string, cred option.ClientOption) (*Storage, error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new client: %s", err)
+	}
+	bkt := client.Bucket(bucketname)
+	return &Storage{bucket: bkt}, nil
+}
+
+// Save uploads the modules .mod, .zip and .info files for a given version
+func (s *Storage) Save(module, version string, mod, zip, info []byte) error {
+	ctx := context.Background()
+	errs := make(chan error, 3)
+	// dispatch go routine for each file to upload
+	go save(ctx, errs, s.bucket, module, version, "mod", mod)
+	go save(ctx, errs, s.bucket, module, version, "zip", zip)
+	go save(ctx, errs, s.bucket, module, version, "info", info)
+
+	errsOut := make([]string, 3)
+	// wait for each routine above to send a value
+	for count := 0; count < 3; count++ {
+		err := <-errs
+		if err != nil {
+			errsOut = append(errsOut, err.Error())
+		}
+	}
+
+	// return concatenated error string if there is anything to report
+	if len(errsOut) < 0 {
+		return fmt.Errorf("one or more errors occured saving %s/@v/%s: %s", module, version, strings.Join(errsOut, ", "))
+	}
+	return nil
+}
+
+// save waits for writeToBucket to complete or times out after five minutes
+func save(ctx context.Context, errs chan<- error, bkt *storage.BucketHandle, module, version, ext string, content []byte) {
+	select {
+	case errs <- writeToBucket(ctx, bkt, fmt.Sprintf("%s/@v/%s.%s", module, version, ext), content):
+		return
+	case <-time.After(5 * time.Minute):
```

Comment:
Shouldn't we define the timeout on the context and use it here? `case <-ctx.Done():`

Reply:
yes indeed, but we need to make sure we'reusing ctx.WithTimeout and not Background one.
also it would be nice to have this value configurable. env should suffice

Reply:
I think overall, `Save` should take a context itself and that can define the timeout

We can totally do that in a separate PR (see #174)

Reply:
i would go with shared timeout. goroutine should start right away but runtime don't give you the guarantee